### DISCUSSION TITLE
logging

### DISCUSSION THREAD

**Discussion by ptabandzelic** [databrickslabs/dbx#443] [Q&A]
Hey, would u recommend  anything for logging.How to configure?How to use inside tasks?Should i create separate logger other than on the spark driver?Any help or advice would be helpfull.
Thanks

**Comment by renardeinside** [comment]
hi @ptabandzelic , 

logging object is already provided by the abstract class `common.Task`. You can use it as `self.logger.info("...")` or other `self.logger.*` methods. 

Logs are visible in the `Driver Logs` section in the UI. Depending on your cloud and setup, it would make sense to export logs to Azure Log Analytics or AWS CloudWatch, or solutions like DataDog etc. Read more on the logging topic [here](https://dbx.readthedocs.io/en/latest/guides/python/python_quickstart/?h=datadog#executing-code-on-databricks).

**Reply by ptabandzelic** [reply]
Yes i know, but can i also configure Spark logger inside Task, also how would u export those logs automatically for example inside dbfs?Or maybe export it somehow inside mlflow run as artifact.I just dont want to go manually and download spark logs.

**Reply by renardeinside** [reply]
> how would u export those logs automatically for example inside dbfs

If it's an interactive cluster, please take a look at Cluster UI -> choose your cluster -> Advanced -> Cluster Logs section -> point logs to specific DBFS or cloud storage. 

> can i also configure Spark logger inside Task

Nope, Spark logging config is configured at the time of cluster start. You can customize it using init scripts, take a look at [this post for details](https://polarpersonal.medium.com/writing-pyspark-logs-in-apache-spark-and-databricks-8590c28d1d51).

**Reply by ptabandzelic** [reply]
Thanks for fast replies.
Its an automated(job) cluster.

**Reply by renardeinside** [reply]
> Its an automated(job) cluster.

In this case you can configure logging properties and parameters by using the `conf/deployment.yml`. 

Add the following lines to the cluster configuration:
```
cluster_log_conf:
    dbfs:
        destination: "dbfs:/some/logs/path"
```

**Reply by ptabandzelic** [reply]
Why i cant use file reference like this and locally make configuration file?
```bash
cluster_log_conf:
  dbfs:
    destination: "file://conf/log4j.properties"

```

**Reply by renardeinside** [reply]
Because it's different files that control different behaviour.
```
cluster_log_conf:
  dbfs:
    destination: "dbfs:/some/logs/path"
```
This setting configures **where** will be logs saved (printed/outputted). 


This file:
> "file://conf/log4j.properties"

Is configuring **how** the logs will be formatted. Therefore this file needs to be provided separately, and set via init script. Please read the blogpost above carefully for details.

**Reply by ptabandzelic** [reply]
Hm, lets be on the same page.
So i want to configure my automated(job) cluster logging. From my understanding i will have to do that with init_scripts inside deployment file like here:  https://kb.databricks.com/clusters/overwrite-log4j-logs.html
So i can provide my init script, locally and file reference it or upload it to dbfs, does not matter. It should just be in deployment file cluster configuration under init_scripts.
What is the point of cluster_log_conf?

**Reply by renardeinside** [reply]
> What is the point of cluster_log_conf?

Cluster log conf defines **where** the logs will be written (try it out - start a job cluster with this setting, then verify the content of this path - you'll see there text files with logs.
